### PR TITLE
Blazor Content Security Policy topic

### LINKED ISSUE #12778: Add guidance on how to enable CSP for Blazor
[EDIT by guardrex to add the topic metadata]

* We looked into enabling this on the templates but deemed it too hard.
* Instead we want to offer some basic guidance on how to use CSP within a project to protect against XSS

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b1a9f84a-feaa-d2b0-409a-01f75f9101a0
* Version Independent ID: aeec639f-a9d6-2f00-b61f-2c316ef6afb6
* Content: [Secure ASP.NET Core Blazor Server apps](https://docs.microsoft.com/en-us/aspnet/core/security/blazor/server?view=aspnetcore-3.0#feedback)
* Content Source: [aspnetcore/security/blazor/server.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/blazor/server.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-security**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #12778

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/security/blazor/content-security-policy?view=aspnetcore-3.1&branch=pr-en-us-17147)

* I didn't see any reason to limit this to Blazor Server, so I provide guidance for Blazor WebAssembly, too. Testing CSP with Blazor WebAssembly went well, but I'll strip that out if there are technical reasons why it can't be supported.
* I qualify the security capability remarks for CSP by saying that CSP "**_helps_**" protect (i.e., no absolute guarantee). Let me know if I was too weak there (or not strong enough).
* Thank you Javier! ... *Your notes, remarks, and suggestions were excellent as usual!* 🎸

### CODE REVIEW

File: aspnetcore/security/blazor/content-security-policy.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+## Policy directives
+
+Minimally, specify the following directives and sources for Blazor apps. Add additional directives and sources as needed. The following directives are used in the [Apply the policy](#apply-the-policy) section of this article, where example security policies for Blazor WebAssembly and Blazor Server are provided:
+
+* [base-uri](https://developer.mozilla.org/docs/Web/HTTP/Headers/Content-Security-Policy/base-uri) &ndash; Restricts the URLs for a document's `<base>` tag.
+  * Specify `self` to indicate that the app's origin, including the scheme and port number, is a valid source.
+  * Supported by all browsers except IE.
+* [block-all-mixed-content](https://developer.mozilla.org/docs/Web/HTTP/Headers/Content-Security-Policy/block-all-mixed-content) &ndash; Prevents loading mixed HTTP and HTTPS content. Supported by all browsers except Edge/IE.
```
Comment (serpent5) [inline aspnetcore/security/blazor/content-security-policy.md]: It looks like this [is supported](https://caniuse.com/#search=block-all-mixed-content) in the re-imagination of Edge. Is it worth being more specific about that?

File: aspnetcore/security/blazor/content-security-policy.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+## Apply the policy
+
+Use a `<meta>` tag to apply the policy:
+
+* Set the value of the `http-equiv` attribute to `Content-Security-Policy`.
+* Place the directives in the `content` attribute value. Separate directives with a semicolon (`;`).
+* Always place the `meta` tag in the `<head>` content.
+
+To test a policy over a period of a few days or weeks without enforcing the policy directives, set `http-equiv` attribute (or header name) to `Content-Security-Policy-Report-Only`. Failure reports are sent as JSON documents to a specified URL. Testing helps confirm that third-party scripts aren't inadvertently blocked when building an initial policy. For more information, see [MDN web docs: Content-Security-Policy-Report-Only](https://developer.mozilla.org/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only).
```
Comment (serpent5) [inline aspnetcore/security/blazor/content-security-policy.md]:
```suggestion
To test a policy over a period of a few days or weeks without enforcing the policy directives, set the `http-equiv` attribute (or header name) to `Content-Security-Policy-Report-Only`. Failure reports are sent as JSON documents to a specified URL. Testing helps confirm that third-party scripts aren't inadvertently blocked when building an initial policy. For more information, see [MDN web docs: Content-Security-Policy-Report-Only](https://developer.mozilla.org/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only).
```

**Comment by blowdart** [comment]
I defer to @javiercn

File: aspnetcore/security/blazor/content-security-policy.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+* [img-src](https://developer.mozilla.org/docs/Web/HTTP/Headers/Content-Security-Policy/img-src) &ndash; Indicates valid sources for images.
+  * Specify `data:` to permit loading images from `data:` URLs.
+  * Specify `https:` to permit loading images from HTTPS endpoints.
+* [object-src](https://developer.mozilla.org/docs/Web/HTTP/Headers/Content-Security-Policy/object-src) &ndash; Indicates valid sources for the `<object>`, `<embed>`, and `<applet>` tags. Specify `none` to prevent all URL sources.
+* [script-src](https://developer.mozilla.org/docs/Web/HTTP/Headers/Content-Security-Policy/script-src) &ndash; Indicates valid sources for scripts.
+  * Specify the `https://stackpath.bootstrapcdn.com/` host source for Bootstrap scripts.
+  * Specify `self` to indicate that the app's origin, including the scheme and port number, is a valid source.
+  * In a Blazor WebAssembly app:
+    * Specify the following hashes to permit inline scripts to load:
```
Comment (mkArtakMSFT) [inline aspnetcore/security/blazor/content-security-policy.md]: It's not clear what these sha values are. I assume these are hashes for the Blazor WASM required script files, so let's be explicit about it.

**Comment by guardrex** [comment]
@mkArtakMSFT Just as well that @javiercn couldn't get to this yet. Sleeping on it *a few days*, I found a number of ... I think 🤔 ... quality updates to apply to the text.

A note on one of the updates ...

I struck the detail on "leaking circuit IDs" because there's no color to the remark. If we keep that, I recommend that we flesh it out ...

* Do they leak because XSS (a script) reveals them to an attacker?
* What's wrong with them leaking? How can they be used to compromise an app? I assume by potentially permitting the attacker to hijack the communication between the client and server.

**Comment by guardrex** [comment]
Scott, I'm fairly comfortable with the textual and layout aspects. This isn't wildly blocking, but it would be nice to get it in quickly so that I have a stable TOC for working more of Javier's security topics into the repo. I'm ready for docs review and will address Javier's remarks when he's available to look.

WRT the title ... I went with "for" (and "with" in the description) over "in" because the framework has nothing to do with CSP. It's *just a header/tag to ignore* as far as Blazor is concerned. "In" didn't seem appropriate for that concept.

**Comment by mkArtakMSFT** [comment]
@guardrex let's get this merged. This looks good. If something comes up later, we'll address then.

**Comment by guardrex** [comment]
We have a community who knows how to stick a 🔪 into a doc author 😨! :smile: I'm sure they'll let me know if any of the guidance 💥 on them.

**Pull request merged**